File: book/HeadFirstDesignPatterns/go/ch8.templateMethod/CaffeineBeverage.go
package main

import "fmt"

type Abstract interface {
	brew()
	addCondiments()
}

type CaffeineBeverage struct {
	Abstract
}

func (c *CaffeineBeverage) PrepareRecipe() {
	c.boilWater()
	c.brew()
	c.pourInCup()
	c.addCondiments()
}

func (c *CaffeineBeverage) boilWater() {
	fmt.Println("물 끓이는 중..")
}

func (c *CaffeineBeverage) pourInCup() {
	fmt.Println("컵에 따르는 중..")
}

File: book/HeadFirstDesignPatterns/java/ch8.template.method/src/CaffeineBeverage.java
public abstract class CaffeineBeverage {

    final void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();
        addCondiments();
        serving();
    }

    abstract void brew();
    abstract void addCondiments();

    void boilWater() {
        System.out.println("물 끓이는 중");
    }

    void pourInCup() {
        System.out.println("컵에 따르는 중");
    }

    void serving() {
        System.out.println("음료가 나왔습니다.");
    }
}
